src/messages_manager.ts

```diff
@@ -12,9 +14,28 @@ export class MessagesManager implements Disposable {
     private disposables: Disposable[] = [];
     private channel: OutputChannel;
 
+    private redrawing = Promise.resolve();
+
+    private revealOutput: boolean = false;
+    private replaceOutput: boolean = false;
+    private displayHistory: boolean = false;
+    private didChange: boolean = false;
+
+    private messageBuffer: string[] = [];
+    private historyBuffer: string[] = [];
+
     public constructor(private readonly main: MainController) {
         this.channel = window.createOutputChannel(`${EXT_NAME} messages`);
-        this.disposables.push(this.channel, eventBus.on("redraw", this.handleRedraw, this));
+
+        // Prevent concurrent redraw / flush by chaining them on a single promise:
+        const redrawHandler = eventBus.on("redraw", (e) => {
+            this.redrawing = this.redrawing.then(() => this.handleRedraw(e));
```

Comment:
I don't think you need this promise. The JS event loop works such that if a `redraw` and `flush-redraw` come in one after the other, one callback will complete execution before the other begins.

Clever way of doing mutal exclusion, though!

Reply:
Is that true? I guess `handleFlush()` will always complete execution, but `handleRedraw()` being async, wouldn't it normally yield at the first `await` point and potentially allow a flush to come in while waiting?

In a previous iteration I had both handlers as `async` so I think it was even more susceptible to this — I would expect execution to long potentially something like this? https://jsfiddle.net/mwboatyk/

Reply:
You're totally right. I had a false memory that this was synchronous (it was when I was last here).